File: ios/Tables/TableCellMeasurementCache.swift
import Foundation

struct TableCellMeasurementKey: Hashable {
    let documentOwner: String
    let contentKey: String
    let innerWidthPixels: Int
    let themeDigest: String
    let fontEnvironmentRevision: Int
    let textScale: CGFloat
    let attachmentRevision: Int
}

final class TableCellMeasurementCache {
    private final class Node {
        let key: TableCellMeasurementKey
        let value: CGFloat
        let bytes: Int
        weak var previous: Node?
        var next: Node?

        init(key: TableCellMeasurementKey, value: CGFloat, bytes: Int) {
            self.key = key
            self.value = value
            self.bytes = bytes
        }
    }

    private var entries: [TableCellMeasurementKey: Node] = [:]
    private(set) var retainedBytes = 0
    private let byteLimit: Int
    private let capacity: Int
    private var mostRecent: Node?
    private var leastRecent: Node?

    var metadataCount: Int { entries.count }

    init(capacity: Int = 32_768, byteLimit: Int? = nil) {
        self.capacity = max(1, capacity)
        self.byteLimit = byteLimit ?? 32 * 1024 * 1024
    }

    func value(for key: TableCellMeasurementKey) -> CGFloat? {
        guard let entry = entries[key] else { return nil }
        moveToMostRecent(entry)
        return entry.value
    }

    func insert(_ value: CGFloat, for key: TableCellMeasurementKey) {
        guard value.isFinite && value >= 0 else { return }
        if let previous = entries.removeValue(forKey: key) {
            unlink(previous)
            retainedBytes -= previous.bytes
        }
        let bytes = max(64, key.documentOwner.utf8.count + key.contentKey.utf8.count + key.themeDigest.utf8.count + 48)
        let entry = Node(key: key, value: value, bytes: bytes)
        entries[key] = entry
        linkAsMostRecent(entry)
        retainedBytes += bytes
        while retainedBytes > byteLimit || entries.count > capacity {
            guard let oldest = leastRecent else { break }
            unlink(oldest)
            entries.removeValue(forKey: oldest.key)
            retainedBytes -= oldest.bytes
        }
    }

    private func moveToMostRecent(_ entry: Node) {
        guard entry !== mostRecent else { return }
        unlink(entry)
        linkAsMostRecent(entry)
    }

    private func linkAsMostRecent(_ entry: Node) {
        entry.previous = nil
        entry.next = mostRecent
        mostRecent?.previous = entry
        mostRecent = entry
        if leastRecent == nil { leastRecent = entry }
    }

    private func unlink(_ entry: Node) {
        let previous = entry.previous
        let next = entry.next
        previous?.next = next
        next?.previous = previous
        if mostRecent === entry { mostRecent = next }
        if leastRecent === entry { leastRecent = previous }
        entry.previous = nil
        entry.next = nil
    }
}

File: android/src/main/java/com/apollohg/editor/tables/TableCellMeasurementCache.kt
package com.apollohg.editor.tables

data class TableCellMeasurementKey(
    val documentOwner: String,
    val contentKey: String,
    val innerWidthPixels: Int,
    val themeDigest: String,
    val fontEnvironmentRevision: Long,
    val textScale: Float,
    val attachmentRevision: Long
)

class TableCellMeasurementCache(private val capacity: Int = 32768, private val byteLimit: Long = 32L * 1024L * 1024L) {
    private data class Entry(val value: Float, val bytes: Long)
    private val entries = LinkedHashMap<TableCellMeasurementKey, Entry>(16, 0.75f, true)
    var retainedBytes: Long = 0
        private set

    fun get(key: TableCellMeasurementKey): Float? = entries[key]?.value

    fun put(key: TableCellMeasurementKey, value: Float) {
        if (!value.isFinite() || value < 0f) return
        entries.remove(key)?.also { retainedBytes -= it.bytes }
        val bytes = maxOf(64L, key.documentOwner.length.toLong() * 2L + key.contentKey.length.toLong() * 2L + key.themeDigest.length.toLong() * 2L + 48L)
        entries[key] = Entry(value, bytes)
        retainedBytes += bytes
        while ((retainedBytes > byteLimit || entries.size > capacity) && entries.isNotEmpty()) {
            val oldest = entries.entries.iterator().next()
            retainedBytes -= oldest.value.bytes
            entries.remove(oldest.key)
        }
    }
}
